### PR TITLE
Support nested workflows by parsing processes recursively

### PR DESCRIPTION
**Pull request by haouech**

### CODE REVIEW

File: taverna-scufl2-cwl/src/main/java/org/apache/taverna/scufl2/cwl/Main.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class Main {
+
+    private static final String HELLO_WORLD_CWL = "/hello_world.cwl";
+    private static final String WORKFLOW_WITH_COMMAND = "/workflow_with_command.cwl";
+    private static JsonNode cwlFile;
+
+    public static void main(String[] args) {
```
Comment (stain) [inline taverna-scufl2-cwl/src/main/java/org/apache/taverna/scufl2/cwl/Main.java]: This sounds like a test class, can we move it to src/test or ideally change it to a junit @Test?

File: taverna-scufl2-cwl/src/main/java/org/apache/taverna/scufl2/cwl/WorkflowParser.java [hunk 2]
```diff
@@ -67,6 +68,12 @@ public WorkflowParser() {
         JsonNode cwlFile = mapper.valueToTree(reader.load(WorkflowParser.class.getResourceAsStream(FILE_NAME)));
         this.cwlParser = new CWLParser(cwlFile);
         this.converter = new Converter();
+        workflowInputs = new HashMap<>();
```
Comment (stain) [inline taverna-scufl2-cwl/src/main/java/org/apache/taverna/scufl2/cwl/WorkflowParser.java]: I don't know why these collections have to be created as part of this constructor, rather move that out to each of their field definitions.

It makes sense to do it in the constructor instead if you have multiple constructors that do it differently, say by passing in values; or if it's complex objects that are intra-connected, as the order of the field constructors is easy to change by accident. But for simple `HashMap` then they can be created in any order as I don't think anything else in the constructor expects them to pre-exist.

File: taverna-scufl2-cwl/src/main/java/org/apache/taverna/scufl2/cwl/components/CommandLineTool.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    private Map<String, InputProcessorPort> processorInputs;
+    private Map<String, OutputProcessorPort> processorOutputs;
+
+    public CommandLineTool(JsonNode node) {
+        this.node = node;
+        this.cwlParser = new CWLParser(node);
+        this.processorInputs = new HashMap<>();
+        this.processorOutputs = new HashMap<>();
+        this.parse();
+        this.receiverPorts = new HashSet(processorInputs.values());
```
Comment (stain) [inline taverna-scufl2-cwl/src/main/java/org/apache/taverna/scufl2/cwl/components/CommandLineTool.java]: `receiverPorts` here is a good example of when the field must be created by the constructor instead. But `processorInputs` and `processorOutputs` should be made outside.

File: taverna-scufl2-cwl/src/main/java/org/apache/taverna/scufl2/cwl/components/CommandLineTool.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        baseCommand = node.get(BASE_COMMAND).asText();
+        parseInputs();
+        parseOutputs();
+    }
+
+    public void parseInputs() {
+        Set<PortDetail> cwlInputs = cwlParser.parseInputs();
+        for(PortDetail detail: cwlInputs) {
+            String portId = detail.getId();
+            InputProcessorPort port = new InputProcessorPort();
```
Comment (stain) [inline taverna-scufl2-cwl/src/main/java/org/apache/taverna/scufl2/cwl/components/CommandLineTool.java]: Add a TODO to set the processor port depth from the CWL type (e.g. checking for `[]`)

File: taverna-scufl2-cwl/src/main/java/org/apache/taverna/scufl2/cwl/components/WorkflowProcess.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+//        writeWorkflowToFile(workflow);
+//
+//        System.out.println("DEBUG WORKFLOW");
+//        System.out.println(workflow.getInputPorts());
+//        System.out.println(workflow.getOutputPorts());
+//        System.out.println(workflow.getProcessors());
+
+    }
+
+    public void writeWorkflowToFile(Workflow workflow) {
```
Comment (stain) [inline taverna-scufl2-cwl/src/main/java/org/apache/taverna/scufl2/cwl/components/WorkflowProcess.java]: This function looks like test code, can you move it to something under src/test?

**Comment by haouech** [comment]
Thank you for your comments.
I added the changes that you requested. I'll work now on adding more unit-tests for the next pull request.

**Comment by stain** [comment]
Sorry, as I saw Jenkins fail this test (for no reason) I merged #40 first. Would you be able to merge against `upstream/master` and fix the conflict?

BTW, Can you delete the empty `Main.java` and move the other `Main.java` to `src/test/main/**` ?

File: taverna-scufl2-cwl/src/test/java/org/apache/taverna/scufl2/cwl/TestWorkflowProcess.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+import org.apache.taverna.scufl2.api.port.InputProcessorPort;
+
+<<<<<<< HEAD
+=======
+import org.apache.taverna.scufl2.api.io.WorkflowBundleIO;
+import org.apache.taverna.scufl2.api.io.WriterException;
+
+import org.apache.taverna.scufl2.api.container.WorkflowBundle;
+
+>>>>>>> 87fa1c18d2b7aa210db3d238905234bf4f52b491
```
Comment (stain) [inline taverna-scufl2-cwl/src/test/java/org/apache/taverna/scufl2/cwl/TestWorkflowProcess.java]: This is trying to submit an unmerged merge conflict. Please revert and try again.

File: taverna-scufl2-cwl/workflow.t2flow.txt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+WorkflowBundle 'bundle1'
```
Comment (stain) [inline taverna-scufl2-cwl/workflow.t2flow.txt]: I don't know if the workflow bundle structure format supports `#` comments for Apache license headers, but if not we will need to add an apache-rat exclude of this file to pom.xml  (there should be some already)

File: taverna-scufl2-cwl/workflow.wfbundle [hunk 8]
```diff
@@ -0,0 +1,10 @@
+WorkflowBundle 'bundle1'
```
Comment (stain) [inline taverna-scufl2-cwl/workflow.wfbundle]: This file has extension .wfbundle but is in text format. Also can we move these to src/test/resources or whatever the files are for?


The file /workflow.vnd.taverna.scufl2.structure.txt` has a weird filename and is not even a text file. Remove.

File: taverna-scufl2-cwl/taverna-scufl2-cwl.iml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (stain) [inline taverna-scufl2-cwl/taverna-scufl2-cwl.iml]: What is this file needed for? Probably should be in .gitignore
